src/mono/wasm/test-main.js

```diff
@@ -243,53 +242,95 @@ const App = {
 };
 globalThis.App = App; // Necessary as System.Runtime.InteropServices.JavaScript.Tests.MarshalTests (among others) call the App.call_test_method directly
 
+function configureRuntime(dotnet, runArgs, INTERNAL) {
+    dotnet
+        .withVirtualWorkingDirectory(runArgs.workingDirectory)
+        .withEnvironmentVariables(runArgs.environmentVariables)
+        .withDiagnosticTracing(runArgs.diagnosticTracing)
+        .withDiagnosticTracing(true)
+        .withExitOnUnhandledError()
+        .withExitCodeLogging()
+        .withElementOnExit();
+
+    if (is_node) {
+        dotnet
+            .withEnvironmentVariable("NodeJSPlatform", process.platform)
+            .withAsyncFlushOnExit();
+
+        const modulesToLoad = runArgs.environmentVariables["NPM_MODULES"];
+        if (modulesToLoad) {
+            dotnet.withModuleConfig({
+                onConfigLoaded: (config) => {
+                    loadNodeModules(config, INTERNAL.require, modulesToLoad)
+                }
+            })
+        }
+    }
+    if (is_browser) {
+        dotnet.withEnvironmentVariable("IsWebSocketSupported", "true");
+    }
+    if (runArgs.runtimeArgs.length > 0) {
+        dotnet.withRuntimeOptions(runArgs.runtimeArgs);
+    }
+    if (runArgs.debugging) {
+        dotnet.withDebugging(-1);
+        dotnet.withWaitingForDebugger(-1);
+    }
+    if (runArgs.forwardConsole) {
+        dotnet.withConsoleForwarding();
+    }
+}
+
+async function dry_run(runArgs) {
+    try {
+        console.log("Silently starting separate runtime instance as another ES6 module to populate caches...");
+        // this separate instance of the ES6 module, in which we just populate the caches
+        const { dotnet, exit, INTERNAL } = await loadDotnet('./dotnet.js?dry_run=true');
+        mono_exit = exit;
+        configureRuntime(dotnet, runArgs, INTERNAL);
+        // silent minimal startup
+        await dotnet.withConfig({
+            forwardConsoleLogsToWS: false,
+            diagnosticTracing: false,
+            appendElementOnExit: false,
+            logExitCode: false,
+            pthreadPoolSize: 0,
+            exitAfterSnapshot: true
+        }).create();
+    } catch (err) {
+        if (err && err.status !== 0) {
+            return false;
+        }
+    }
+    console.log("Separate runtime instance finished loading.");
+    return true;
+}
+
 async function run() {
     try {
+        const runArgs = await getArgs();
+        console.log("Application arguments: " + runArgs.applicationArguments.join(' '));
+
+        if (is_browser) {
+            const dryOk = await dry_run(runArgs);
+            if (!dryOk) {
+                mono_exit(1, "Failed during dry run");
+                return;
+            }
+        }
+
+        // this is subsequent run with the actual tests. It will use whatever was cached in the previous run. 
+        // This way, we are testing that the cached version works.
```

Comment:
I'm fine with this in principle, but it does seem like down the road we could run into issues where all of our automated tests are relying on the behavior of a loaded snapshot, and we end up with first-run-only bugs. I'm not sure if there's anything you can do about that though.

Reply:
There  are lot of tests running on V8 and NodeJS. They don't get this change.